### PR TITLE
Data Update Script to migrate logo_svg contents to png file (Will be merged and deployed on 10/01)

### PR DESCRIPTION
**Pull request by Ridhwana**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
As per https://forem.team/engineering/creator-signup-and-onboarding-16p6, we've updated the logo upload process for creators as follows:
- We only accept PNG/JPG/JPEG to be uploaded as a logo.
- We no longer accept SVG due to security vulnerabilities.

Based on the above we need to convert existing Forem’s logos from SVG (i.e, the `logo_svg` markup) to PNG so that there is no action for Forem Creators. 

Please note that this is a best effort script and we try to cater for as many conditions/edge cases as possible but we are unable to cater for everything. Hence we will be giving creators the opportunity to change/tweak their logos and re-upload if the resolution, visibility has been altered during the conversion. In addition, we will remind them that they should view their logo on both dark and light theme.

There are some Forems (very few it seems, like DEV, forem.dev and vscodetips) who have added custom scripts or are using dynamic CSS variables in their SVG logo to alter the logo’s styles for our dark theme and light theme (an example is the DEV logo). We are not able to convert these logos accurately with our tools.

Hence, we do our best effort to first provide static values for these CSS variable (that we know will not work in dark theme) and for others we want this script to log the errors for any of the logos that fail to convert.

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/15620

## QA Instructions, Screenshots, Recordings

_Test a successful update:_ 
1. Open a rails console and set the value for `logo_svg` 
```
Settings::General.logo_svg = '<svg width="300" height="100" xmlns="http://www.w3.org/2000/svg">
    <path fill="#dedede" stroke="#555" stroke-width="2" d="M2 2h296v96H2z"/>
    <text x="50%" y="50%" font-size="18" text-anchor="middle" alignment-baseline="middle" font-family="monospace, sans-serif" fill="#555">
        300×100
    </text>
</svg>'
```

2. Make sure that the logo's are nil. Open a rails console and run:
```
Settings::General.original_logo = nil 
Settings::General.resized_logo = nil
```
3. Run the Data Update Script in your terminal with `data_updates:run`
4. Open the rails console and check that the values for `Settings::General.original_logo` and `Settings::General.resized_logo` have updated.
5. Go to `/public/uploads/logos` and you should see an original and resized png versions of the original svg. 

You can also test a logo with a css variable like `currentColor` or `var(--base-inverted)`  and watch it convert if you'd like.
### UI accessibility concerns?

None.

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help 

Some insights into fixing the Travis tests:
Initially the MiniMagick conversion process is failing on Travis.
```
CarrierWave::ProcessingError:
       Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /home/travis/build/forem/forem/tmp/1640166684-774651839101422-2549-8379/300x100.svg -auto-orient /tmp/image_processing20211222-11351-73ymxc.png` failed with error:
       convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
       convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
     # ./spec/uploaders/logo_svg_uploader_spec.rb:39:in `block (3 levels) in <main>'
     # ./spec/[secure]_helper.rb:145:in `block (3 levels) in <main>'
     # ./spec/[secure]_helper.rb:145:in `block (2 levels) in <main>'
     # ./spec/[secure]_helper.rb:100:in `block (2 levels) in <main>'
     # ------------------
     # --- Caused by: ---
     # MiniMagick::Error:
     #   `convert /home/travis/build/forem/forem/tmp/1640166684-774651839101422-2549-8379/300x100.svg -auto-orient /tmp/image_processing20211222-11351-73ymxc.png` failed with error:
     #   convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
     #   convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
     #   ./spec/uploaders/logo_svg_uploader_spec.rb:39:in `block (3 levels) in <main>'

```

 I tried a couple of things thus far like:
- Making sure that the fixture file is found as per [Mac's comment](https://github.com/forem/forem/pull/15710#discussion_r773195512)
- Testing the functionality with the fixture file locally (I moved it temporarily into the uploads folder): 
```
svg_file = File.open('./public/uploads/300x100.svg')
logo_svg_uploader = LogoSvgUploader.new.tap do |uploader|
  uploader.store!(svg_file)
end
```
 and it worked.
- https://travis-ci.community/t/build-times-out-with-no-apparent-reason/5083 as per this [commit](https://github.com/forem/forem/pull/15710/commits/53d96d006af2b7b6ab90344a336c294abe54639e).
- https://stackoverflow.com/questions/41138404/how-to-install-newer-imagemagick-with-webp-support-in-travis-ci-container as per this [commit](https://github.com/forem/forem/pull/15710/commits/9835d466e66e0eebfdfcdfc0b889b40acc3b3453)

I suspect it's a Travis configuration issue or maybe version difference in Travis, but haven't been able to figure it out as yet and would appreciate any insights on it.

See more of the conversation here: https://forem-team.slack.com/archives/C7GE84DEJ/p1641305439053700

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

I've chatted to Jennie about alerting Forems about the following: 
Forems will have a week to preview their converted uploads to check for:
1. resolution 
2. visibility 
3. and determine if they'd like to twek/update their logo to work on both dark adn light theme

Pawel will be tweaking our logos as well. 

We will send out a forem.dev post once we have run the script.

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

File: app/uploaders/base_uploader.rb
Comment (Ridhwana) [inline app/uploaders/base_uploader.rb:29]: There will be no exif data for an SVG

File: app/uploaders/logo_svg_uploader.rb
Comment (Ridhwana) [inline app/uploaders/logo_svg_uploader.rb:11]: We set a process callback, which will call that method any time a file is uploaded. Thereafter the MiniMagick convert method which changes the image encoding format to the given format.

File: app/uploaders/logo_uploader.rb
Comment (Ridhwana) [inline app/uploaders/logo_uploader.rb:32]: This is a bit of an unrelated change but part of the same feature, we're not using the `original_filename` as part of the new file name, hence there is no need to check for it.

File: lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb
Comment (Ridhwana) [inline lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:9]: We create a temporary file with the SVG string

File: lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb
Comment (Ridhwana) [inline lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:20]: This is to ensure that we are logging the error, so that we can alert the Forems whose SVG's were not able to convert successfully.

File: lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb
Comment (Ridhwana) [inline lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:22]: (blocking): I'm not sure how to use the ForemStatsClient properly, any idea's here?

File: spec/uploaders/logo_svg_uploader_spec.rb
Comment (Ridhwana) [inline spec/uploaders/logo_svg_uploader_spec.rb:37]: (blocking): These tests are failing on Travis with a MiniMagick error: 
```
Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /home/travis/build/forem/forem/tmp/1639126885-827978499212382-0004-7279/300x100.svg -auto-orient /tmp/image_processing20211210-10742-7n7kz8.png` failed with error:
       convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
       convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
```

but passing locally.

Comment (maestromac) [reply spec/uploaders/logo_svg_uploader_spec.rb:37]: Hmm, try changing the following context on rails helper
```
-  config.fixture_path = "#{::Rails.root}/spec/fixtures"
+  config.fixture_path = "#{::Rails.root}/spec/fixtures/files"
```

File: lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb
Comment (maestromac) [reply lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:22]: That's our metric wrapper. This line of code would keep count on how often `logo_svg.conversion` action happens. It would result in a graph format that can be seen on Datadog.

File: app/uploaders/base_uploader.rb
Comment (jeremyf) [reply app/uploaders/base_uploader.rb:29]: (nonblocker): This seems like a good inline comment for the code.

File: lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb
Comment (citizen428) [reply lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:9]: The ["Good practices" sectiob](https://ruby-doc.org/stdlib-3.0.2/libdoc/tempfile/rdoc/Tempfile.html#class-Tempfile-label-Good+practices) of the `Tempfile` documentation encourages the following form instead:

```ruby
Tempfile.create("logo.svg") do |file|
   # ...do something with file... e.g.
  file.write(::Settings::General.logo_svg))
end
```

To quote the [docs](https://ruby-doc.org/stdlib-3.0.2/libdoc/tempfile/rdoc/Tempfile.html#method-c-open)

> The File object will be automatically closed and the temporary file is removed after the block terminates, releasing all resources that the block created. The call returns the value of the block.

**Review by citizen428 — APPROVED** [review]

File: .travis.yml
Comment (Ridhwana) [inline .travis.yml:77]: I butchered the Travis configuration, but I would like to set a specific version of ImageMagick as per this [comment](https://stackoverflow.com/questions/41138404/how-to-install-newer-imagemagick-with-webp-support-in-travis-ci-container) and see if that helps.

**Comment by djuber** [comment]
testing with a debug build suggests the errors in convert/minimagick/carrierwave are resolved by installing the `gsfonts` package in the image, before running rpsec.

https://forem-team.slack.com/archives/C7GE84DEJ/p1641317978064000?thread_ts=1641305439.053700&cid=C7GE84DEJ

File: .travis.yml
Comment (Ridhwana) [inline .travis.yml:50]: We needed to add this so that the Travis build would pass. Thank you Dan!

File: app/uploaders/base_uploader.rb
Comment (Ridhwana) [reply app/uploaders/base_uploader.rb:30]: ✅

File: lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb
Comment (Ridhwana) [inline lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:23]: I would love a second pair of eyes on this from someone that is familiar with Honeybadger. 

We are expecting some logo conversions to fail and we want to be able to easily identify the failures in Honeybadger and notify the Forems that they need to do the conversion manually. Hence I've added the context and then a tag to easily filter these errors.

**Comment by Ridhwana** [comment]
@citizen428 I made some changes and hence re-requesting your review.

File: app/uploaders/logo_svg_uploader.rb
Comment (jeremyf) [inline app/uploaders/logo_svg_uploader.rb:45]: (non-blocking) Do we want to memoize this?  Wondering if we want `original_logo_asdfasdf.png` and `resized_logo_asdfasdf.png`?

File: lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb
Comment (jeremyf) [reply lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:23]: @jgaskins might be a good reviewer of this

File: app/uploaders/logo_svg_uploader.rb
Comment (Ridhwana) [reply app/uploaders/logo_svg_uploader.rb:45]: I mean I don't think it would hurt, but maybe not a priority since this most likely a once-off script that we're running for existing Forems to migrate their existing logos. 

I could go in and optimise this on LogoUploader as that is used for our normal upload process, I'll go in and do this in a subsequent PR.

File: lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb
Comment (Ridhwana) [reply lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:23]: Thanks Jeremy, I think Jamie is on PTO but I've requested his review as well.

File: lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb
Comment (Ridhwana) [inline lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:13]: Please note that this is a best effort script and we try to cater for as many cases where the user may have hacked their logo with CSS variable to work differently within the light and dark theme. These css variables will fall away and are converted to static values. 

However, we are unable to cater for all cases and we just replacing the most common css variables. 

We will be giving creators the opportunity to change/tweak their logos before launch.

**Review by jeremyf — APPROVED** [review]

**Review by msarit — COMMENTED** [review]
@Ridhwana Hey there!

The happy path works as expected, but the DataUpdateScript for the failure path is succeeding (when it should fail) 🤔  I was sure to follow your excellent (🎉 ) QA instructions.

**Comment by Ridhwana** [comment]
> @Ridhwana Hey there!
> 
> The happy path works as expected, but the DataUpdateScript for the failure path is succeeding (when it should fail) 🤔 I was sure to follow your excellent (🎉 ) QA instructions.

Oops.... I forgot to update the QA instructions when I added some [best effort updates to help convert logos that convert CSS variables](https://github.com/forem/forem/pull/15710#discussion_r779644145). Sorry Arit!
With the update referenced above (i.e. replacing the css variables with a static value)  we no longer need to test this as it will pass like you observed.

**Review by msarit — APPROVED** [review]

File: app/uploaders/logo_svg_uploader.rb
Comment (maestromac) [inline app/uploaders/logo_svg_uploader.rb:45]: Did you mean to put this comment on line 39 instead?

**Review by maestromac — APPROVED** [review]
:clap: worked beautifully locally!

But to reiterate, once merged this job will definitely run on DEV and it will definitely fail?

File: lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb
Comment (jgaskins) [inline lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:35]: Only suggestion here is to move `tags` outside of `context` since `tags` is a [top-level property](https://www.rubydoc.info/gems/honeybadger/Honeybadger/Agent#notify-instance_method) of the Honeybadger notice. That should make it easier to search for.

```suggestion
      Honeybadger.notify(e, context: {
                           community_name: Settings::Community.community_name,
                           app_domain: Settings::General.app_domain,
                         },
                         tags: "failed_svg_conversion")
```

**Review by jgaskins — APPROVED** [review]
Just one suggestion. Beyond that, this looks great!

The search you'll want to use on Honeybadger is [this one](https://app.honeybadger.io/projects/72638/faults?sort=last_seen_desc&q=tag%3A%22failed_svg_conversion%22%E2%80%81-is%3Aresolved%E2%80%81-is%3Aignored%E2%80%81). Also, thank you so much for your attention there! :100:

**Review by juliannatetreault — APPROVED** [review]
Let's do this! 🚀

**Review by citizen428 — APPROVED** [review]

**Comment by Ridhwana** [comment]
> 👏 worked beautifully locally!
> 
> But to reiterate, once merged this job will definitely run on DEV and it will definitely fail?

@maestromac this should not fail on DEV locally anymore, and I've tested this. We changed track based on this [curveball](https://forem-team.slack.com/archives/C01S4QZUDNC/p1641461142000400?thread_ts=1641403412.444700&cid=C01S4QZUDNC) and we added [some functionality to replace the common CSS variables with static values](https://github.com/forem/forem/pull/15710/files#r779644145). This will help Forems like DEV.to and Forem.dev to pass. If all goes well with this change we expect most Forems to pass because after looking at the data it seems to be mostly our Forems that use CSS Variables (which is now catered for in the script).

**Review by nickytonline — APPROVED** [review]
🚀

**Pull request merged**
